src/npg_notify/ont/event.py

```diff
@@ -0,0 +1,463 @@
+        return cls(**serializable)
+
+
+def add_email_tasks(
+    pipeline: Pipeline, event: EventType, reader, writer
+) -> tuple[int, int, int]:
+    """Add new ONT event email tasks to the pipeline.
+
+    For each collection read from the reader, a new task is created and added to the
+    pipeline. The collection is then written to the writer.
+
+    Args:
+        pipeline: The pipeline to which the tasks are added.
+        event: The event that triggers the email.
+        reader: A reader of ONT run collections in baton JSON format, one per line.
+           These collections must have the standard ONT run metadata:
+              - ont:experiment_name
+              - ont:instrument_slot
+              - ont:flowcell_id
+        writer: A writer to which the collections are written after processing, also
+           in baton JSON format, one per line.
+    Returns:
+        The number of collections processed, the number of tasks added, and the number
+        of errors encountered.
+    """
+    np = ns = ne = 0
+
+    for line in reader:
+        np += 1
+        try:
+            coll = Collection.from_json(line)
+            try:
+                expt = coll.avu("ont:experiment_name").value
+                slot = int(coll.avu("ont:instrument_slot").value)
+                flowcell_id = coll.avu("ont:flowcell_id").value
+            except ValueError as e:
+                raise MetadataError(
+                    "Collection does not have the metadata required for ONT event email"
+                ) from e
+
+            task = ContactEmail(
+                experiment_name=expt,
+                instrument_slot=slot,
+                flowcell_id=flowcell_id,
+                path=coll.path.as_posix(),
+                event=event,
+            )
+            if pipeline.add(task):
+                ns += 1
+                log.info("Task added", pipeline=pipeline, task=task)
+            else:
+                log.info("Task already exists", pipeline=pipeline, task=task)
+
+            print(coll.to_json(indent=None, sort_keys=True), file=writer)
+        except Exception as e:
+            ne += 1
+            log.exception(e)
+
+    return np, ns, ne
+
+
+def run_email_tasks(
+    pipeline: Pipeline, session: Session, domain: str
+) -> tuple[int, int, int]:
+    """Claim tasks from the pipeline and run them to send ONT event emails.
+
+    For each task (representing an ONT run) an email will be sent to all the contacts
+    of the studies associated with the run.
+
+    Args:
+        pipeline: The pipeline whose tasks are to be run.
+        session: An open MLWH DB session.
+        domain: A network domain name to use when sending email. The main will be sent
+           from mail.<domain> with <user>@<domain> in the From: header.
+
+    Returns:
+        The number of tasks processed, the number of tasks that succeeded, and the
+        number of errors encountered.
+    """
+    np = ns = ne = 0
+    batch_size = 1
```

Comment:
The batch size should be increased to the maximum number of emails that it's reasonable to send at one time